cpmpy/solvers/pindakaas.py

```diff
@@ -0,0 +1,322 @@
+        """
+        if isinstance(cpm_var, NegBoolView):  # negative literal
+            # get inner variable and return its negated solver var
+            return ~self.solver_var(cpm_var._bv)
+        elif isinstance(cpm_var, _BoolVarImpl):  # positive literal
+            # insert if new
+            if cpm_var.name not in self._varmap:
+                self._varmap[cpm_var.name] = self.pkd_solver.add_variable()
+            return self._varmap[cpm_var.name]
+        elif isinstance(cpm_var, _IntVarImpl):
+            raise NotSupportedError(
+                f"{self.name}: integer variable {cpm_var} of type {type(cpm_var)} not supported"
+            )
+        else:
+            raise NotImplementedError(
+                f"{self.name}: unexpected variable {cpm_var} of type {type(cpm_var)} not"
+            )
+
+    def transform(self, cpm_expr):
+        """
+            Transform arbitrary CPMpy expressions to constraints the solver supports.
+
+            Implemented through chaining multiple solver-independent **transformation functions** from
+            the `cpmpy/transformations/` directory.
+
+            See the 'Adding a new solver' docs on readthedocs for more information.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        cpm_cons = toplevel_list(cpm_expr)
+        cpm_cons = decompose_in_tree(cpm_cons)
+        cpm_cons = simplify_boolean(cpm_cons)
+        cpm_cons = flatten_constraint(cpm_cons)  # flat normal form
+        cpm_cons = only_bv_reifies(cpm_cons)
+        cpm_cons = only_implies(cpm_cons)
+        cpm_cons = linearize_constraint(
+            cpm_cons, supported=frozenset({"sum", "wsum", "and", "or"})
+        )
+        return cpm_cons
+
+    def __add__(self, cpm_expr_orig):
+        """
+        Eagerly add a constraint to the underlying solver.
+
+        Any CPMpy expression given is immediately transformed (through `transform()`)
+        and then posted to the solver in this function.
+
+        This can raise 'NotImplementedError' for any constraint not supported after transformation
+
+        The variables used in expressions given to add are stored as 'user variables'. Those are the only ones
+        the user knows and cares about (and will be populated with a value after solve). All other variables
+        are auxiliary variables created by transformations.
+
+        What 'supported' means depends on the solver capabilities, and in effect on what transformations
+        are applied in `transform()`.
+
+        """
+        import pindakaas as pkd
+
+        if self.unsatisfiable:
+            return self
+
+        # add new user vars to the set
+        get_variables(cpm_expr_orig, collect=self.user_vars)
+
+        # transform and post the constraints
+        try:
+            for cpm_expr in self.transform(cpm_expr_orig):
+                self._add(cpm_expr)
```

Comment:
is this worth putting in a separate function, if it is only used here?
(function calls can be expensive in hot paths in Python... no inlineing)

Reply:
This function is used later to handle the implication, which makes it much cleaner (compare to say how pysat.py handles implications).

For performance, recursion can be worse, but in this case we are guaruantueed the deph is at most one (due to linearize). In other cases with more depth, I would still do it like this and use profiling. I'm very in favour of this style of recursively traversing expression trees, because it makes the code much much cleaner. If your function handles `e : Expr`, and this happens to be `p : BV -> sub : Expr`, then you should call the function again on sub. You will see the same pattern in int2bool.